saleor/payment/utils.py

```diff
@@ -125,9 +158,102 @@ def clean_authorize(payment: Payment):
         raise PaymentError('Charged transactions cannot be authorized again.')
 
 
+def call_gateway(
+        func_name, transaction_kind, payment, payment_token, **extra_params):
+    """Helper that calls the passed gateway function and handles exceptions.
+
+    Additionally does validation of the returned gateway response."""
+    gateway, gateway_params = get_payment_gateway(payment.gateway)
+    gateway_response = None
+    error_msg = None
+
+    payment_information = create_payment_information(
+        payment, payment_token, **extra_params
+    )
+
+    try:
+        gateway_response = getattr(gateway, func_name)(
+            payment_information=payment_information, **gateway_params)
+        validate_gateway_response(gateway_response)
+    except AttributeError:
+        error_msg = 'Gateway doesn\'t implement {}'.format(func_name)
+        logger.exception(error_msg)
+    except GatewayError:
+        error_msg = 'Gateway response validation failed'
+        logger.exception(error_msg)
+        gateway_response = None  # set response empty as the validation failed
+    except Exception:
+        error_msg = 'Gateway encountered an error'
+        logger.exception(error_msg)
+    finally:
+        if not isinstance(gateway_response, list):
+            gateway_response = [gateway_response]
+        transactions = []
+        for response in gateway_response:
+            transactions.append(create_transaction(
+                payment=payment,
+                kind=transaction_kind,
+                payment_information=payment_information,
+                error_msg=error_msg,
+                gateway_response=response))
+
+    for transaction in transactions:
+        if not transaction.is_success:
+            # attempt to get errors from response, if none raise a generic one
+            raise PaymentError(transaction.error or GENERIC_TRANSACTION_ERROR)
+
+    return transactions[-1]
+
+
+def validate_gateway_response(responses):
```

Comment:
Do we though? Technically that does not make the response invalid, and I think if the gateway made a successful payment in a different currency (because of reasons) we still want to keep it as a successful one instead of raising an error that marks it as unsuccessful. Perhaps log a warning if the returned currency is indeed different. What do you think?

Reply:
Actually that is quite important, for example in https://github.com/mirumee/saleor/blob/3eca2f4de9bef55a8ec3ea3fcda22024995bb0a6/saleor/payment/utils.py#L190, the payment's captured amount is updated from the transaction's amount directly without checking the currency. Saleor currently only support default currency, to be honest. And then the captured amount is used to determine if a order is fully paid. So correct amount with same currency is important.

If a gateway's returned currency is not same as saleor's, maybe there will be a problem. For example, saleor may think it charges 10USD, but the gateway may charges 10EURO. But saleor calculate captured amount without considering the currency. 

But in this case, the gateway made a successfull payment indeed. Maybe it is not proper to mark it as unsuccessfully since the money is actually charged though the currency is incorrect. I agree with you, maybe let's just add warnings for now.

Reply:
Added a warning 👍